theano/tensor/signal/downsample.py

```diff
@@ -268,78 +268,123 @@ def grad(self, inp, grads):
+          ||(PyArray_DIMS(%(z)s)[2] != z_r)
+          ||(PyArray_DIMS(%(z)s)[3] != z_c)
           )
         {
           if (%(z)s) Py_XDECREF(%(z)s);
           npy_intp dims[4] = {0,0,0,0};
           dims[0]=PyArray_DIMS(%(x)s)[0];
           dims[1]=PyArray_DIMS(%(x)s)[1];
-          dims[2]=z_shp0;
-          dims[3]=z_shp1;
+          dims[2]=z_r;
+          dims[3]=z_c;
           //TODO: zeros not necessary
           %(z)s = (PyArrayObject*) PyArray_ZEROS(4, dims, typenum,0);
         }
 
-        if (z_shp0 && z_shp1)
+        // used for indexing a pool region inside the input
+        int r_st, r_end, c_st, c_end;
+        int maximum; // temp var for maximum value in a region 
+        if (z_r && z_c)
         {
-            for(int b=0;b<PyArray_DIMS(%(x)s)[0];b++){
-              for(int k=0;k<PyArray_DIMS(%(x)s)[1];k++){
-                int mini_i = 0;
-                int zi = 0;
-                for(int i=0;i< x_shp0_usable; i++){
-                  int mini_j = 0;
-                  int zj = 0;
-                  for(int j=0; j<x_shp1_usable; j++){
-                    dtype_%(x)s a = ((dtype_%(x)s*)(PyArray_GETPTR4(%(x)s,b,k,i,j)))[0];
-                    dtype_%(z)s * __restrict__ z = ((dtype_%(z)s*)(PyArray_GETPTR4(%(z)s,b,k,zi,zj)));
-                    z[0] = (((mini_j|mini_i) == 0) || z[0] < a) ? a : z[0];
-                    mini_j = ((mini_j + 1) == %(ds1)s) ? 0 : mini_j+1;
-                    zj += (mini_j == 0);
+            for(int b=0; b<PyArray_DIMS(%(x)s)[0]; b++){
+              for(int k=0; k<PyArray_DIMS(%(x)s)[1]; k++){
+                for(int i=0; i< z_r; i++){
+                  r_st = i * %(st0)s;
+                  r_end = r_st + %(ds0)s;
+                  // skip the padding
+                  r_st = (r_st < %(pd0)s) ? %(pd0)s : r_st;
+                  r_end = (r_end > (r + %(pd0)s)) ? r + %(pd0)s : r_end;
+                  //printf("r_st %%d, r_end %%d",r_st, r_end);
+                  for(int j=0; j<z_c; j++){
+                    c_st = j * %(st1)s;
+                    c_end = c_st + %(ds1)s;
+                    // skip the padding
+                    c_st = (c_st < %(pd1)s) ? %(pd1)s : c_st;
+                    c_end = (c_end > (c + %(pd1)s)) ? c + %(pd1)s : c_end;
+                    //printf("c_st %%d, c_end %%d ",c_st, c_end);
+                    dtype_%(z)s * z = (
+                          (dtype_%(z)s*)(PyArray_GETPTR4(%(z)s, b, k, i, j)));
+                    // use the first element  as the initial value of maximum
+                    maximum = ((dtype_%(x)s*)(PyArray_GETPTR4(%(x)s,b,k,r_st,c_st)))[0];
+                    printf("maximum= %%.5f ",maximum);
+                    // go through the pooled region in the input
+                    for(int m=r_st; m<r_end; m++)
+                    {
+                      for(int n=c_st; n<c_end; n++)
+                      {
+                        dtype_%(x)s a = ((dtype_%(x)s*)(PyArray_GETPTR4(%(x)s,b,k,m,n)))[0];
```

Comment:
I don't understand what you mean. For the first patch (pooling region), for instance, `i = 0`, `j = 0`, so `r_st = 0` and `c_st = 0`. The first pooling region starts at the corner of the padded image, so these indices refer to the corner of the padded image.
Then, you take the maximum of `c_st` and `%(pd1)s`, to have the effective start point, which is the coordinate of the original image inside the padded one. Until now it is consistent: you are using coordinates where 0 mean the start of the padded image, and the first pixel is at `(%(pd0)s, %(pd1)s)`.
The problem is that you then try to access `x[b, k, r_st, c_st]` first, which is `x[b, k, pd0, pd1]`, which is not the first pixel of the original image. You want to access `x[b, k, 0, 0]` instead.